### PR TITLE
auth: Reintroduce GSS-TSIG processing

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Based on the code that was removed in #9385 with the following changes and additions:

- Even when compiled in, the accepting is only done if the new setting `enable-gss-tsig` is set to `yes`.
- We only try to create a `GssContext` if we succeeded in finding an accepting principal to avoid creating security contexts that are useless.
- The three maps used by `GssContext` are now thread safe.
- The three maps have an expiry mechanism
- If the update query fails, e.g. because we could not find a matching`TSIG-ALLOW-DNSUPDATE` metadata, we immediately clean up the associated security context after the reply has been sent.
- There are regression tests.

A slightly related change is that if the reading the `lua-dnsupdate-policy-script` fails, no Lua policy is activated.

Also contains improvements to the auth startup and teardown code in the `auth-py` regression test, copied from the rec regression tests.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [X] included documentation (including possible behaviour changes)
- [X] documented the code
- [X] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: .github/actions/spell-check/expect.txt [hunk 1]
```diff
@@ -1861,6 +1862,7 @@ Winfried
 wireformat
 wirelength
 Wisiol
+wmem
 Wmissing
```
Comment (jsoref) [inline .github/actions/spell-check/expect.txt:1866]: Fwiw, from https://github.com/check-spelling/check-spelling/wiki/Configuration-Examples:-patterns, one could probably adapt some subset of this:
```
# Compiler flags
[\t "'`=]-[DPWXYLlf]
```

To drop `Wmissing` (not sure it's worth it).

**Comment by omoerbeek** [comment]
Failing tests are dependent on external servers, alas. Ignore for the moment.

**Comment by omoerbeek** [comment]
Rebased

**Comment by jsoref** [comment]
Please cherry-pick #11467 onto this branch

**Comment by Habbie** [comment]
> Please cherry-pick #11467

Huh? Please do not do this.

**Comment by Habbie** [comment]
> Huh? Please do not do this.

I forgot that a later rebase will get rid of that commit again, if it has appeared on master meanwhile. So, such cherry-picks are okay if they get rid of problems, as long as we then do that rebase later, and make sure the cherry-picked commit has disappeared.

File: docs/tsig.rst [hunk 2]
```diff
@@ -131,10 +129,12 @@ using Kerberos with TSIG signatures.
...
 
--  Working Kerberos environment. Please refer to your Kerberos vendor
-   documentation on how to setup it.
--  Principal (such as ``DNS/<your.dns.server.name>@REALM``) in either
-   per-user keytab or system keytab.
+-  Working Kerberos environment. Please refer to your Kerberos vendor documentation on how to set it up.
+-  Accepting Principal (of the form ``DNS/your.dns.server.name@REALM``) in either per-user keytab or
+   system keytab, where ``your.dns.server.name`` must match the nameserver name in the SOA record of the zone.
+   If a user keytab is used, specify it using the ``KRB5_KTNAME`` environment variable when starting up PDNS server,
+   which must be able to read the keytab file.
```
Comment (Habbie) [inline docs/tsig.rst:136]: please format as one sentence per line

File: pdns/Makefile.am [hunk 3]
```diff
@@ -225,6 +229,7 @@ pdns_server_SOURCES = \
 	ednsoptions.cc ednsoptions.hh \
 	ednssubnet.cc ednssubnet.hh \
 	histogram.hh \
+        gss_context.cc gss_context.hh \
```
Comment (Habbie) [inline pdns/Makefile.am:232]: indent mismatch

File: pdns/gss_context.cc [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+    if (d_peer_name != GSS_C_NO_NAME) {
+      tmp_maj = gss_release_name(&tmp_min, &(d_peer_name));
+    }
+  }
+
+  std::shared_ptr<GssCredential> d_cred;
+  GssContextType d_type{GSS_CONTEXT_NONE};
+  gss_ctx_id_t d_ctx{GSS_C_NO_CONTEXT};
+  gss_name_t d_peer_name{GSS_C_NO_NAME};
+  time_t d_expires{time(nullptr) + 60}; // partly initialized wil be cleaned up
```
Comment (Habbie) [inline pdns/gss_context.cc:168]: is this different from `d_expires = 0` in the constructor on purpose?

File: pdns/tkey.cc [hunk 5]
```diff
@@ -2,11 +2,19 @@
...
 #endif
 #include "packethandler.hh"
+#include "gss_context.hh"
+#include "common_startup.hh"
 
 void PacketHandler::tkeyHandler(const DNSPacket& p, std::unique_ptr<DNSPacket>& r) {
+#if 0
+  auto [i,a,s] = GssContext::getCounts();
+  cerr << "#init_creds: " << i << " #accept_creds: " << a << " #secctxs: " << s << endl;
+#endif
```
Comment (Habbie) [inline pdns/tkey.cc:12]: how about always logging this at Debug?

File: regression-tests.auth-py/kerberos-client/init-keytab.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+echo commands to run:
+echo Passwords enterd shoudl match those in kerberos-server setup script
```
Comment (Habbie) [inline regression-tests.auth-py/kerberos-client/init-keytab.sh:2]:
```suggestion
echo Passwords entered should match those in the kerberos-server setup script
```

**Review by Habbie — CHANGES_REQUESTED** [review]
Awesome work! I posted a few questions/suggestions.

File: pdns/rfc2136handler.cc [hunk 7]
```diff
@@ -683,10 +683,20 @@ int PacketHandler::processUpdate(DNSPacket& p) {
...
+        GssName inputname(p.d_peer_principal); // match against principal since GSS
+        for(const auto& key: tsigKeys) {
+          if (inputname.match(key)) {
+            validKey = true;
+            break;
+          }
+        }
+      } else {
+        for(const auto& key: tsigKeys) {
+          if (inputkey == DNSName(key)) { // because checkForCorrectTSIG has already been performed earlier on, if the names of the ky match with the domain given. THis is valid.
```
Comment (cmouse) [inline pdns/rfc2136handler.cc:698]: `THis` => `This.`

File: pdns/rfc2136handler.cc [hunk 8]
```diff
@@ -683,10 +683,20 @@ int PacketHandler::processUpdate(DNSPacket& p) {
         return RCode::Refused;
       }
 
-      for(const auto& key: tsigKeys) {
-        if (inputkey == DNSName(key)) { // because checkForCorrectTSIG has already been performed earlier on, if the names of the ky match with the domain given. THis is valid.
-          validKey=true;
-          break;
+      if (p.d_tsig_algo == TSIG_GSS) {
+        GssName inputname(p.d_peer_principal); // match against principal since GSS
```
Comment (cmouse) [inline pdns/rfc2136handler.cc:689]: since GSS what?

File: pdns/gss_context.cc [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+
+  recv_tok.length = input.size();
+  recv_tok.value = (void*)input.c_str();
+  sign_tok.length = signature.size();
+  sign_tok.value = (void*)signature.c_str();
+
+  maj = gss_verify_mic(&min, d_ctx->d_ctx, &recv_tok, &sign_tok, NULL);
+
+  if (maj != GSS_S_COMPLETE) {
+    processError("gss_get_mic", maj,min);
```
Comment (cmouse) [inline pdns/gss_context.cc]: missing space after maj,

File: pdns/gss_context.cc [hunk 10]
```diff
@@ -40,58 +40,66 @@
...
-  ~GssCredential() {
-    OM_uint32 tmp_maj  __attribute__((unused)), tmp_min __attribute__((unused));
-    if (d_cred != GSS_C_NO_CREDENTIAL) 
+  ~GssCredential()
+  {
+    OM_uint32 tmp_maj __attribute__((unused)), tmp_min __attribute__((unused));
+    if (d_cred != GSS_C_NO_CREDENTIAL)
       tmp_maj = gss_release_cred(&tmp_min, &d_cred);
-    if (d_name != GSS_C_NO_NAME) 
+    if (d_name != GSS_C_NO_NAME)
```
Comment (cmouse) [inline pdns/gss_context.cc]: wonder if it would make more sense to just (void) instead of tmp_maj?

File: pdns/gss_context.cc [hunk 11]
```diff
@@ -94,9 +95,9 @@ class GssCredential : boost::noncopyable
   bool renew()
   {
     OM_uint32 time_rec, tmp_maj __attribute__((unused)), tmp_min __attribute__((unused));
-    d_maj = gss_acquire_cred(&d_min, d_name, GSS_C_INDEFINITE, GSS_C_NO_OID_SET, d_usage, &d_cred, nullptr, &time_rec);
+    tmp_maj = gss_acquire_cred(&tmp_min, d_name, GSS_C_INDEFINITE, GSS_C_NO_OID_SET, d_usage, &d_cred, nullptr, &time_rec);
```
Comment (cmouse) [inline pdns/gss_context.cc:100]: tmp_maj is no longer unused.

File: pdns/gss_context.cc [hunk 12]
```diff
@@ -194,7 +194,7 @@ static void expire()
 {
   static time_t s_last_expired;
   time_t now = time(nullptr);
-  if (now - s_last_expired < 5) {
+  if (now - s_last_expired < 60) {
```
Comment (cmouse) [inline pdns/gss_context.cc:199]:
```
#define TSIG_GSS_EXPIRE_TIMEOUT 60
```
?

File: pdns/distributor.hh [hunk 13]
```diff
@@ -249,7 +249,9 @@ retry:
 
       QD->callback(a, QD->start);
 #ifdef ENABLE_GSS_TSIG
-      QD->Q.cleanupGSS(a->d.rcode);
+      if (a != nullptr) {
```
Comment (cmouse) [inline pdns/distributor.hh:252]: why not put this in the commit where this was already introduced?

File: docs/settings.rst [hunk 14]
```diff
@@ -649,6 +649,17 @@ This setting MUST be 32 hexadecimal characters, as the siphash algorithm's key u
...
 
+.. _setting-enable-gss-tsig:
+
+``enable-gss-tsig``
+-------------------
+
+-  Boolean
+-  Default: no
+
+Enable the accepting of GSS-TSIG signed messages.
```
Comment (cmouse) [inline docs/settings.rst:660]: `Enable accepting GSS-TSIG signed messages`

File: pdns/tcpreceiver.cc [hunk 15]
```diff
@@ -627,14 +646,15 @@ int TCPNameserver::doAXFR(const DNSName &target, std::unique_ptr<DNSPacket>& q,
     DNSName algorithm=trc.d_algoName; // FIXME400: check
```
Comment (cmouse) [inline pdns/tcpreceiver.cc:646]: old fixme.

File: pdns/gss_context.cc [hunk 16]
```diff
@@ -0,0 +1,588 @@
...
+  bool valid()
+  {
+    return d_valid && !expired();
+  }
+
+  std::string d_nameS;
+  gss_cred_usage_t d_usage;
+  gss_name_t d_name{GSS_C_NO_NAME};
+  gss_cred_id_t d_cred{GSS_C_NO_CREDENTIAL};
+  time_t d_expires{time(nullptr) + 60}; // partly initialized wil be cleaned up
```
Comment (cmouse) [inline pdns/gss_context.cc:129]: `partly initialized will be cleaned up`

File: pdns/gss_context.cc (hunk 4 above)
Comment (omoerbeek) [reply pdns/gss_context.cc:164]: This was a bug indeed. I cganged the constructor to not overwrite `d_expires` (and use the init values for the other fields as well).

File: pdns/tkey.cc (hunk 5 above)
Comment (omoerbeek) [reply pdns/tkey.cc:12]: Done

File: docs/tsig.rst [hunk 17]
```diff
@@ -130,26 +130,21 @@ Prerequisites
 ~~~~~~~~~~~~~
 
 -  Working Kerberos environment. Please refer to your Kerberos vendor documentation on how to set it up.
--  Accepting Principal (of the form ``DNS/your.dns.server.name@REALM``) in either per-user keytab or
-   system keytab, where ``your.dns.server.name`` must match the nameserver name in the SOA record of the zone.
-   If a user keytab is used, specify it using the ``KRB5_KTNAME`` environment variable when starting up PDNS server,
-   which must be able to read the keytab file.
+-  Accepting Principal (of the form ``DNS/your.dns.server.name@REALM``) in either per-user keytab or system keytab, where ``your.dns.server.name`` must match the nameserver name in the SOA record of the zone.
```
Comment (jsoref) [inline docs/tsig.rst:133]:
```suggestion
-  Accepting Principals (of the form ``DNS/your.dns.server.name@REALM``) in either per-user keytab or system keytab, where ``your.dns.server.name`` must match the nameserver name in the SOA record of the zone.
```

**Comment by omoerbeek** [comment]
Squashed. Go to merge once green as far as I am concerned.

File: pdns/dnspacket.cc [hunk 18]
```diff
@@ -756,3 +759,14 @@ bool DNSPacket::checkForCorrectTSIG(UeberBackend* B, DNSName* keyname, string* s
 const DNSName& DNSPacket::getTSIGKeyname() const {
   return d_tsigkeyname;
 }
+
+#ifdef ENABLE_GSS_TSIG
+void DNSPacket::cleanupGSS(int rcode)
+{
+  if (rcode != RCode::NoError && d_tsig_algo == TSIG_GSS && !getTSIGKeyname().empty()) {
```
Comment (Habbie) [inline pdns/dnspacket.cc:766]: should this also check if the feature is enabled globally? The downstream `erase` is, of course, very cheap, so perhaps it's fine like this.

**Review by Habbie — COMMENTED** [review]
One nit, that may apply in multiple places. Rephrased as a question: do we check `g_doGssTSIG` in enough places?

File: pdns/dnspacket.cc (hunk 18 above)
Comment (omoerbeek) [reply pdns/dnspacket.cc:766]: I will review all the spots where checking `g_doGssTSIG` might be relevant.

**Comment by omoerbeek** [comment]
Local builds were ok, but CI not, will investigate.

**Review by Habbie — APPROVED** [review]

File: regression-tests.auth-py/test_GSSTSIG.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+
+    def testNoAcceptor(self):
+        self.kinit("testuser1")
+        self.nsupdate("add inserted12.noacceptor.net 10 A 1.2.3.12", 2)
+        self.checkNotInDB('example.net', 'inserted12.example.net')
+
+    def testWrongAcceptor(self):
+        self.kinit("testuser1")
+        self.nsupdate("add inserted13.wrongacceptor.net 10 A 1.2.3.13", 2)
+        self.checkNotInDB('example.net', 'inserted13.example.net')
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply regression-tests.auth-py/test_GSSTSIG.py:134]: Yes, looks like a copy-paste error. Will fix soon.

[the remark this answers is not in the document]
Comment (omoerbeek) [reply regression-tests.auth-py/test_GSSTSIG.py:134]: https://github.com/PowerDNS/pdns/pull/13745

**Pull request merged**